### ISSUE TITLE
Changes that should be inconsequential result in crashes and inconsistencies when a thread is involved.

### ISSUE THREAD

**Issue by BigZaphod** [godotengine/godot#75873]
### Godot version

v4.0.2.stable.official [7a0977ce2]

### System information

Apple M1 Max, MacBook Pro 16-inch, 2021, 64 GB, macOS Ventura 13.3 (22E252)

### Issue description

I have encountered a situation where simply reading (not changing) some constants from multiple threads, it appears to completely and totally change program behavior to the point where it will randomly crash or corrupt the on-screen results.

I was working on a Minecraft clone and I started migrating my chunk generation to threads when I started having all manner of difficulties with crashes and corrupted meshes. At first I assumed I was doing something wrong, but as I tried to isolate the problem, I kept deleting code and removing elements and found myself becoming more and more confused and now I'm am reasonably confident that I must have found a bug. During this process I uncovered three separate and seemingly innocuous code changes that could each trigger trouble for me and have documented them below and demonstrated them in the attached video.

Since this involves multithreading, I would not be surprised if it turns out to be difficult to reproduce it, unfortunately. I have only tested this on my mac, so there's a chance it doesn't even happen on Windows/Linux, but it occurs repeatedly and reliably for me as you will see.

I am using the standard release of Godot 4.0.2 as downloaded from https://godotengine.org. I do not know if this is new in this version as I only just encountered it about 3 days ago and I have spent all of my free time since pulling everything I could out of my project to narrow down and isolate a reasonably small reproducible (at least for me) case.

### Steps to reproduce

The exact steps to reproduce (at least for me) is to download and open my sample project and simply try to run it a few times. You should see that it crashes or doesn't always display properly. Sometimes it just hangs. Other times it causes macOS to popup the standard "app crashed" dialog box.

I put together a video to demonstrate all of this, but Github says it's too big so I split it up into parts.

Here's what the program should look like when it's running correctly:

https://user-images.githubusercontent.com/453721/230800010-6435a1ba-d6c2-464a-bc37-9fc2c2ab9e9f.mov

Now, here's some short snippets going through the different ways to trigger the problems.

There are at least three very simple bits of code that trigger these behaviors as I tried to show in the videos.

The first is in `Global.gd`. This script is installed as an Autoload named `Global`. As demonstrated in the video below, the `BLOCKS` array is initially this:

`var BLOCKS := [ .. etc ..]`

However, simply changing the `var` to `const` will cause the program to crash:

https://user-images.githubusercontent.com/453721/230800057-c85295f6-d03f-4ce7-aaf4-6685014c4739.mov

The crash is usually shown as one of the lines in the `update_geometry` function, but that's not always the case. As you can see in the video, the crash even moves around - the line that's blamed changes each time I run it and sometimes it manages to launch, but you can see the visuals are incorrect.

When I change `const` back to `var`, the program once against seems to work fine:

https://user-images.githubusercontent.com/453721/230800072-68e1f442-ba97-4865-b7c5-de74b4272107.mov

The next change is, IMO, even stranger - simply changing the name of the enum constant `SOLID` will also change the results! Even more bizarrely, `SOLID` isn't even being used in the program at all!

https://user-images.githubusercontent.com/453721/230800089-28ca6c37-aeb5-47f4-9cd7-a0f6dfff8b8f.mov

What's totally wild is simply changing `SOLID` to some other name will *also* work! It just doesn't like the name `SOLID`, I guess? I can even leave the `BLOCKS` array declared as `const` and the program will still work:

https://user-images.githubusercontent.com/453721/230800111-3f1db90e-e0d8-4dc8-a96e-5ef48ed300a5.mov

As you can see, changing things back to the way they were before will once again trigger the crashing:

https://user-images.githubusercontent.com/453721/230800141-94566aa6-1a5d-4887-a140-9ef39b5e4fcf.mov

And the final trigger I found is to simply comment out an *unused* line of code in the `set_block_id()` function. If I remove that line, then I can leave `SOLID` in place and leave the `BLOCKS` array as `const` and the program somehow still works:

https://user-images.githubusercontent.com/453721/230800185-153fa259-8428-414a-ad79-fd64fae864a6.mov

And once again, simply putting the line back - even though it isn't used for anything - results in crashing again:

https://user-images.githubusercontent.com/453721/230800216-71d6b1a3-e407-4f36-b6a3-c431ea976a99.mov

I'm reasonably baffled by all of this.

When the program first starts, it attempts to create 4 "chunks" of terrain using a single worker thread which is hardcoded to only make these four chunks - nothing especially fancy is going on here and I've removed tons of stuff to get a simple reproduction. Here's all of `Main.gd`:

```gdscript
extends Node3D

const ChunkScene := preload("res://Chunk.tscn")

var thread := Thread.new()

func _ready() -> void:
	thread.start(thread_function)

func thread_function() -> void:
	for chunk_position in [Vector3i(0, 0, 0), Vector3i(0, 0, -1), Vector3i(0, 0, 1), Vector3i(1, 0, 0), Vector3i(-1, 0, 0)]:
		var chunk := make_chunk(chunk_position)
		back_on_main.call_deferred(chunk_position, chunk)

func back_on_main(chunk_position: Vector3i, chunk: Chunk) -> void:
	chunk.position = chunk_position * Global.CHUNK_SIZE
	add_child(chunk)

func make_chunk(chunk_position: Vector3i) -> Chunk:
	var chunk_origin: Vector3i = chunk_position * Global.CHUNK_SIZE
	var chunk: Chunk = ChunkScene.instantiate()

	var noise := FastNoiseLite.new()
	noise.frequency = 0.0025
	noise.fractal_octaves = 10
	noise.seed = 0

	for z in Global.CHUNK_SIZE.z:
		for x in Global.CHUNK_SIZE.x:
			var global_x := chunk_origin.x + x
			var global_z := chunk_origin.z + z
			var height := ceili(((noise.get_noise_2d(global_x, global_z) + 1) / 2) * 192)
				
			for y in range(0, height):
				var block_id := 1

				if y >= height - 5:
					block_id = 2

				chunk.set_block_id(Vector3i(x, y, z), block_id)
				
	return chunk
```

Each chunk is an instance of a chunk scene and the blocks are all set before it is added to the scene tree. The adding happens with a `call_deferred` call so it runs on the main thread as Godot expects and after that, the chunk instances aren't touched by any other threads as far as I'm aware.

The `main_chunk()` function isn't very complicated - it just uses noise to make a terrain. In my testing, this doesn't seem to matter very much - adding a bunch of hardcoded blocks also seems to break things, so I don't think using noise has anything to do with anything.

The code in `Chunk.gd` is a bit more complicated in places and is pretty dumb in this example, but that's because I removed as much as I could from my bigger project to build this reproduction. It is likely that most of this isn't necessary to trigger the bug(s), but it does help make it super easy to see when the bug occurs so I left it here.

I hesitate to even guess what's causing this, but at a high level it almost has to be some kind of memory corruption - simply changing the name of `SOLID` to `SOLID1` really shouldn't have changed anything, and yet it clearly does. My initial assumption was that `const` vs. `var` might be the primary culprit, but then other changes seem to negate that. The very confusing case of simply not calling that one kind of code in `set_block_id()` suggests that maybe accessing a dictionary has some kind of subtle role to play in all this?

In any case... I really wish someone luck in figuring this one out and I really hope someone else will be able to reproduce it.

### Minimal reproduction project

[GodotThreadingBug.zip](https://github.com/godotengine/godot/files/11186439/GodotThreadingBug.zip)

**Comment by asmaloney** [comment]
Can reproduce using:

8-Core Intel Core i9, iMac 27-inch, 2019, 32 GB, macOS 12.6.4

Godot 4.0.2 built from source.

Here's the stack:
```
Engine version: Godot Engine v4.0.2.stable.custom_build (7a0977ce2c558fe6219f0a14f8bd4d05aea8f019)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] 1   libsystem_platform.dylib            0x00007ff8082f7dfd _sigtramp + 29
[2] 2   ???                                 0x00000000066b95d3 0x0 + 107714003
[3] HashMap<Variant, Variant, VariantHasher, StringLikeVariantComparator, DefaultTypedAllocator<HashMapElement<Variant, Variant> > >::find(Variant const&) (in godot.macos.editor.dev.x86_64) (hash_map.h:467)
[4] Dictionary::getptr(Variant const&) const (in godot.macos.editor.dev.x86_64) (dictionary.cpp:108)
[5] VariantKeyedSetGetDictionary::get(Variant const*, Variant const*, Variant*, bool*) (in godot.macos.editor.dev.x86_64) (variant_setget.cpp:987)
[6] Variant::get_keyed(Variant const&, bool&) const (in godot.macos.editor.dev.x86_64) (variant_setget.cpp:1153)
[7] Variant::get(Variant const&, bool*) const (in godot.macos.editor.dev.x86_64) (variant_setget.cpp:1205)
[8] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Callable::CallError&, GDScriptFunction::CallState*) (in godot.macos.editor.dev.x86_64) (gdscript_vm.cpp:948)
[9] GDScriptInstance::callp(StringName const&, Variant const**, int, Callable::CallError&) (in godot.macos.editor.dev.x86_64) (gdscript.cpp:1828)
[10] Object::callp(StringName const&, Variant const**, int, Callable::CallError&) (in godot.macos.editor.dev.x86_64) (object.cpp:711)
[11] Variant::callp(StringName const&, Variant const**, int, Variant&, Callable::CallError&) (in godot.macos.editor.dev.x86_64) (variant_call.cpp:1161)
[12] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Callable::CallError&, GDScriptFunction::CallState*) (in godot.macos.editor.dev.x86_64) (gdscript_vm.cpp:1666)
[13] GDScriptInstance::callp(StringName const&, Variant const**, int, Callable::CallError&) (in godot.macos.editor.dev.x86_64) (gdscript.cpp:1828)
[14] bool Node::_gdvirtual__process_call<false>(double) (in godot.macos.editor.dev.x86_64) (node.h:238)
[15] Node::_notification(int) (in godot.macos.editor.dev.x86_64) (node.cpp:58)
[16] Node::_notificationv(int, bool) (in godot.macos.editor.dev.x86_64) (node.h:46)
[17] Node3D::_notificationv(int, bool) (in godot.macos.editor.dev.x86_64) (node_3d.h:0)
[18] VisualInstance3D::_notificationv(int, bool) (in godot.macos.editor.dev.x86_64) (visual_instance_3d.h:0)
[19] GeometryInstance3D::_notificationv(int, bool) (in godot.macos.editor.dev.x86_64) (visual_instance_3d.h:0)
[20] MeshInstance3D::_notificationv(int, bool) (in godot.macos.editor.dev.x86_64) (mesh_instance_3d.h:0)
[21] Object::notification(int, bool) (in godot.macos.editor.dev.x86_64) (object.cpp:790)
[22] SceneTree::_notify_group_pause(StringName const&, int) (in godot.macos.editor.dev.x86_64) (scene_tree.cpp:874)
[23] SceneTree::process(double) (in godot.macos.editor.dev.x86_64) (scene_tree.cpp:466)
[24] Main::iteration() (in godot.macos.editor.dev.x86_64) (main.cpp:3169)
[25] OS_MacOS::run() (in godot.macos.editor.dev.x86_64) (os_macos.mm:668)
[26] main (in godot.macos.editor.dev.x86_64) (godot_main_macos.mm:77)
```

**Comment by asmaloney** [comment]
Some additional info after playing around a bit. I started outputting the block id and colour in Chunk's `update_geometry` and removed the `add_face_geometry` calls:

```gdscript
var idx := (z * Global.CHUNK_SIZE.x * Global.CHUNK_SIZE.y) + (y * Global.CHUNK_SIZE.x) + (x)
				
if idx >= blocks.size():
	push_error("Index out of range")
	continue
					
var block_id := blocks[idx]
if block_id > 0:
	var block = Global.BLOCKS[block_id]
	var pos := Vector3i(x, y, z)
	var colour:Color = block[Global.COLOR]
	var format_str = "({y}, {x}, {z}) {id} {colour}"
	print(format_str.format({"y": y, "x":x, "z":z, "id":block_id, "colour": colour}))
```

I'm getting this:

```
E 0:00:01:0040   Chunk.gd:48 @ update_geometry(): Trying to unreference a SafeRefCount which is already zero is wrong and a symptom of it being misused.
Upon a SafeRefCount reaching zero any object whose lifetime is tied to it, as well as the ref count itself, must be destroyed.
Moreover, to guarantee that, no multiple threads should be racing to do the final unreferencing to zero.
  <C++ Error>    FATAL: Condition "count.get() == 0" is true.
  <C++ Source>   ./core/templates/safe_refcount.h:176 @ _check_unref_sanity()
  <Stack Trace>  Chunk.gd:48 @ update_geometry()
                 Chunk.gd:27 @ _process()
```

Line 48 is:
```gdscript
var block = Global.BLOCKS[block_id]
```

It feels a bit like a [Heisenbug](https://en.wikipedia.org/wiki/Heisenbug) because I don't get that message all the time and it depends on the code around it. Maybe something's not initialized properly?

**Comment by BigZaphod** [comment]
> It feels a bit like a [Heisenbug](https://en.wikipedia.org/wiki/Heisenbug) because I don't get that message all the time and it depends on the code around it. Maybe something's not initialized properly?

I'm completely unfamiliar with the internals of Godot, but perhaps since the array is declared `const` some code somewhere is not setting up an initial retain count of `1` on the array and/or dictionaries that it contains or something like that? (Although if that's the case, it seems like problems should occur all over the place and not just when using a thread, I'd think. 🤔)

I also wondered if accessing `BLOCKS` through the `Global` Autoload somehow was a factor here, but after moving `BLOCKS` to be a local const in the `Chunk.gd` file, it didn't seem to make any difference.

**Comment by BigZaphod** [comment]
I just realized something I didn't notice before: when I commented out `SOLID` and the program then worked... that should not have been possible. I completely forgot that the unused line of code I reference later actually *does* use the `SOLID` enum - and yet... it didn't complain when running or even crash when the `SOLID` enum itself was actually commented out despite that line still being there?

The line in `set_block_id()` that I comment/uncomment in the last videos:

```gdscript
var _is_solid = Global.BLOCKS[block_id].get(Global.SOLID, true)
```

It clearly uses `Global.SOLID` there.... wtf?!

**Comment by BigZaphod** [comment]
Ok I think I have a theory about part of this that might narrow some things down and I suspect there are at least two bugs in play. The first is some kind of issue with reading from a `const` array (and/or dictionary) from multiple threads at the same time - something that should and is supposed to be possible as far as I can tell. The second is that GDScript errors that occur off the main thread are being swallowed up and ignored somewhere.

For this second issue, as I noted in my previous message, `Global.SOLID` was actually being used and I had just forgotten that when I was going through my initial reproduction steps and recording of the video clips - it was used in `set_block_id()` in the discarded line of code. What has occurred to me now is that when renaming or commenting out `SOLID`, the program behaves the same way as when I commented out this line in `set_block_id()`:

```gdscript
var _is_solid = Global.BLOCKS[block_id].get(Global.SOLID, true)
```

So, effectively, the program is continuing as if I had commented that line out when I comment out or rename `SOLID`.

It occurred to me that GDScript doesn't discover the undefined `Global.SOLID` symbol in this case until it tries to *use* it. So to test this theory, I commented out `SOLID` as before and removed threading by changing `Main.gd`'s `_ready()` function to this:

```gdscript
func _ready() -> void:
	#thread.start(thread_function)
	thread_function()
```

This just calls the `thread_function()` on the normal main thread instead. When I run the program with `SOLID` commented out, this immediately results in an obvious GDScript error at runtime on the line in `set_block_id()`:

<img width="350" alt="image" src="https://user-images.githubusercontent.com/453721/231525934-15868aac-c561-47b0-bf92-cd0cc5ca26ba.png">

So far so good.

Now, I changes `set_block_id()` to put this unused line of code *before* setting the `block_id` in the array:

```gdscript
func set_block_id(pos: Vector3i, block_id: int) -> void:
	var idx := (pos.z * Global.CHUNK_SIZE.x * Global.CHUNK_SIZE.y) + (pos.y * Global.CHUNK_SIZE.x) + (pos.x)

	# STRANGE THING:
	# Commenting out this line appears to stop the crashing/corruption?!
	var _is_solid = Global.BLOCKS[block_id].get(Global.SOLID, true)

	blocks[idx] = block_id
```

Then I re-enabled threading in Main and re-ran it. Now the program once again launches just fine, but with no visible blocks:

<img width="1264" alt="image" src="https://user-images.githubusercontent.com/453721/231526732-baf754be-3db8-4e19-8910-bd21db020e4e.png">

My theory here is that when GDScript throws the exception about `Global.SOLID` not existing, something is swallowing it up and ignoring it. The function is aborted, but execution continues. There is no indication in the editor or console that an error occurred because of the missing `Global.SOLID` symbol at all.

So that is one thing explained - there's another Godot bug or limitation here where GDScript errors in a thread are just totally swallowed up and ignored. Most likely that should get a new issue filed, I guess?

The main threading bug remains present, however, but I think maybe this narrows it down to `const` arrays/dictionary in particular having a problem when being read from by multiple threads. It's curious that a `var` array doesn't seem to suffer from this, so my hypothesis is that perhaps some code somewhere made an assumption that `const` arrays don't need to setup a lock or they are setting it up incorrectly in a way that just happens to break when two things access it in parallel.

**Comment by BigZaphod** [comment]
So it seems that what really triggers this is the presence of that unused line in `set_block_id()`. Likely, specifically, accessing the `Global.BLOCKS` array and/or accessing the dictionary element within that array using `get()` at the same moment that the main thread is also accessing one of those same objects in the `update_geometry()` function that is being run by a prior generated chunk.

**Comment by dweipert-3138720606** [comment]
The same error that happens with `const` also happens with `static var`. Using a regular var from a `Global` autoload works around this currently as well.
